curvefs/src/client/warmup/warmup_manager.cpp

```diff
@@ -0,0 +1,702 @@
+                }
+                if (pendingReq.fetch_sub(1, std::memory_order_seq_cst) == 1) {
+                    VLOG(6) << "pendingReq is over";
+                    cond.Signal();
+                }
+                delete[] context->buf;
+                return;
+            }
+            if (++context->retry >= downloadMaxRetryTimes_) {
+                if (pendingReq.fetch_sub(1, std::memory_order_seq_cst) == 1) {
+                    VLOG(6) << "pendingReq is over";
+                    cond.Signal();
+                }
+                VLOG(9) << "Up to max retry times, "
+                        << "download object failed, key: " << context->key;
+                delete[] context->buf;
+                return;
+            }
+
+            LOG(WARNING) << "Get Object failed, key: " << context->key
+                         << ", offset: " << context->offset;
+            s3Adaptor_->GetS3Client()->DownloadAsync(context);
+        };
+
+    pendingReq.fetch_add(prefetchObjs.size(), std::memory_order_seq_cst);
+    if (pendingReq.load(std::memory_order_seq_cst)) {
+        VLOG(9) << "wait for pendingReq";
+        for (auto iter : prefetchObjs) {
+            VLOG(9) << "download start: " << iter.first;
+            std::string name = iter.first;
+            uint64_t readLen = iter.second;
+            if (s3Adaptor_->GetDiskCacheManager()->IsCached(name)) {
+                pendingReq.fetch_sub(1);
+                continue;
+            }
+            char *cacheS3 = new char[readLen];
+            memset(cacheS3, 0, readLen);
+            auto context = std::make_shared<GetObjectAsyncContext>();
+            context->key = name;
+            context->buf = cacheS3;
+            context->offset = 0;
+            context->len = readLen;
+            context->cb = cb;
+            context->retry = 0;
+            s3Adaptor_->GetS3Client()->DownloadAsync(context);
+        }
+        if (pendingReq.load())
+            cond.Wait();
+    }
+}
+
+bool WarmupManagerS3Impl::ProgressDone(fuse_ino_t key) {
+    bool ret;
+    {
+        ReadLockGuard lockList(warmupFilelistDequeMutex_);
+        ret = LockedFindKeyWarmupFilelist(key) == warmupFilelistDeque_.end();
+    }
+
+    inode2FetchDentryPoolMutex_.RDLock();
+    ret = ret &&
+          (LockedFindKeyFetchDentryPool(key) == inode2FetchDentryPool_.end());
+    inode2FetchDentryPoolMutex_.Unlock();
+
+    {
+        ReadLockGuard lockInodes(warmupInodesDequeMutex_);
+        ret =
+            ret && (LockedFindKeyWarmupInodes(key) == warmupInodesDeque_.end());
+    }
+
+
+    inode2FetchS3ObjectsPoolMutex_.RDLock();
+    ret = ret && (LockedFindKeyFetchS3ObjectsPool(key) ==
+                  inode2FetchS3ObjectsPool_.end());
+    inode2FetchS3ObjectsPoolMutex_.Unlock();
+    return ret;
+}
+
+void WarmupManagerS3Impl::ScanCleanFetchDentryPool() {
+    // clean inode2FetchDentryPool_
+    inode2FetchDentryPoolMutex_.RDLock();
```

Comment:
> Can we just use `WRLock` instead of interleaving reader and writer lock?

Will the stop time be longer? , so an alternate approach would allow data to be inserted?